File: app/src/main/java/com/ducktapedapps/updoot/model/CommentData.java
package com.ducktapedapps.updoot.model;

import java.io.Serializable;

public class CommentData implements Data, Serializable {
    private String author;
    private int depth;
    private String body;
    private int ups;
    private Boolean likes;
    private boolean saved;
    private String id;
    private Thing replies;
    private Gildings gildings;

    public String getAuthor() {
        return author;
    }

    public int getDepth() {
        return depth;
    }

    public String getBody() {
        return body;
    }

    public int getUps() {
        return ups;
    }

    public Boolean getLikes() {
        return likes;
    }

    public boolean isSaved() {
        return saved;
    }

    public String getId() {
        return id;
    }

    public Thing getReplies() {
        return replies;
    }

    @Override
    public String toString() {
        return "CommentData{" +
                "author='" + author + '\'' +
                ", depth=" + depth +
                ", body='" + body + '\'' +
                ", ups=" + ups +
                ", likes=" + likes +
                ", saved=" + saved +
                ", id='" + id + '\'' +
                ", replies=" + replies +
                ", Gildings=" + gildings +
                '}';
    }

    public Gildings getGildings() {
        return gildings;
    }
}

File: app/src/main/java/com/ducktapedapps/updoot/model/CommentData.kt
package com.ducktapedapps.updoot.model

import android.text.Spanned

data class CommentData(
        override val id: String,
        override val depth: Int,
        override val parent_id: String,
        override val name: String,
        val author: String,
        var body: Spanned,
        var ups: Int?,
        val likes: Boolean?,
        val replies: List<BaseComment>,
        val gildings: Gildings,
        val repliesExpanded: Boolean = false,
        val is_submitter: Boolean,
        val author_flair_text: String
) : BaseComment(id, depth, name, parent_id) {

    fun vote(direction: Int): CommentData {
        var updatedLikes: Boolean? = this.likes
        var updatedUps = this.ups
        when (direction) {
            1 -> if (this.likes == null) {
                updatedLikes = true
                if (updatedUps != null)
                    updatedUps++
            } else if (!this.likes) {
                updatedLikes = true
                if (updatedUps != null)
                    updatedUps += 2
            } else {
                updatedLikes = null
                if (updatedUps != null)
                    updatedUps--
            }
            -1 -> when {
                this.likes == null -> {
                    if (updatedUps != null)
                        updatedUps--
                    updatedLikes = false
                }
                this.likes -> {
                    if (updatedUps != null)
                        updatedUps -= 2
                    updatedLikes = false
                }
                else -> {
                    if (updatedUps != null)
                        updatedUps++
                    updatedLikes = null
                }
            }
        }
        return this.copy(
                ups = updatedUps,
                likes = updatedLikes
        )
    }
}